### PR TITLE
feat: add authentication endpoints

### PR DESCRIPTION
**Pull request by MonicaTanios**
* create database structure migration
 * add register service/endpoint
 * add login service/endpoint
 * create register/login dto

Co-authored-by: Micheline Medhat <michelinemedhat@gmail.com>

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,6 +4,7 @@
 # compiled output
 /dist
 /node_modules
+/prisma/migrations
```
Comment (yousinix) [inline .gitignore]: Migrations should be checked in source control

File: package.json [hunk 2]
```diff
@@ -21,7 +21,8 @@
     "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
     "test:e2e": "jest --config ./test/jest-e2e.json",
     "db:gen": "prisma generate",
-    "db:push": "prisma db push --preview-feature --accept-data-loss"
+    "db:push": "prisma db push --preview-feature --accept-data-loss",
+    "db:migrate": "prisma migrate dev --name init"
```
Comment (yousinix) [inline package.json]: Don't add `--name ...`

File: package.json [hunk 3]
```diff
@@ -21,7 +21,8 @@
     "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
     "test:e2e": "jest --config ./test/jest-e2e.json",
     "db:gen": "prisma generate",
-    "db:push": "prisma db push --preview-feature --accept-data-loss"
+    "db:push": "prisma db push --preview-feature --accept-data-loss",
```
Comment (yousinix) [inline package.json]: Replace this script by one that pushes migrations to db

File: prisma/schema.prisma [hunk 4]
```diff
@@ -6,9 +6,15 @@ datasource db {
...
-  email String  @unique
-  name  String?
+  id            String    @id
+  email         String    @unique
+  phone_number  String    @unique
+  full_name     String
+  password      String
+  credit        Float     @default(0)
+  seed          String    @db.VarChar(6)
+  secret        String
```
Comment (yousinix) [inline prisma/schema.prisma]: `seed` and `secret` are the same thing as far as I can understand, aren't they?

File: src/users/users.service.ts [hunk 5]
```diff
@@ -1,39 +1,67 @@
...
-    return this.users.find((u) => u.id == id);
+  async findOne(id: string): Promise<User> {
+    const user = await this.prisma.user.findUnique({
+      where: {
+        id: id,
+      },
+    });
+    if (user === null) {
+      throw new NotFoundException('Incorrect user credentials.');
+    }
```
Comment (yousinix) [inline src/users/users.service.ts]: Don't throw exception, just return the `user` which would be `null` if not found.

File: src/users/users.service.ts [hunk 6]
```diff
@@ -1,39 +1,67 @@
...
+import { LoginUserDto } from './dto/login-user.dto';
 
 @Injectable()
 export class UsersService {
-  private readonly users: User[] = [];
-
   constructor(private readonly prisma: PrismaService) {}
 
-  create(): User {
+  async register(registerUserDto: RegisterUserDto): Promise<User> {
```
Comment (yousinix) [inline src/users/users.service.ts]: This should be create as it was, the register/login logic should be handled in the controller.

File: src/users/users.service.ts [hunk 7]
```diff
@@ -1,39 +1,67 @@
...
+        data: {
+          ...registerUserDto,
+          secret: encryptedSecret,
+        },
+      });
+    } catch (err) {
+      if (err instanceof Prisma.PrismaClientKnownRequestError) {
+        throw new BadRequestException(
+          `There is a unique constraint violation on ${err.meta['target']} field.`,
+        );
```
Comment (yousinix) [inline src/users/users.service.ts]: The thrown exception should have the following properties:

- Status code: [`409`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409)
- `validationErrors` property 

See [`ValidationException`](https://github.com/tazkrtak/server/blob/master/src/infrastructure/validation-exception.ts), I suggest creating another exception called `NonUniqueException` with the above properties.

File: src/users/users.controller.ts [hunk 8]
```diff
@@ -1,19 +1,30 @@
 import { ApiCreatedResponse, ApiOperation, ApiTags } from '@nestjs/swagger';
-import { Controller, Get } from '@nestjs/common';
-import { User } from './interfaces/user.interface';
+import { Body, Controller, Get, Post } from '@nestjs/common';
 import { UsersService } from './users.service';
+import { RegisterUserDto } from './dto/register-user.dto';
+import { User } from './interfaces/user.interface';
```
Comment (yousinix) [inline src/users/users.controller.ts]: Rename the `User` to `UserDto` because it will conflict with prisma's types.

**Review by yousinix — CHANGES_REQUESTED** [review]
@MichaelSafwatHanna, please review the DTO's validation part _(using prisma as we discussed before)_. You can check the comments too because I am not fully aware of the server project.

File: src/users/dto/register-user.dto.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  phone_number: string;
+
+  @IsString()
+  full_name: string;
+
+  @IsString()
+  password: string;
+
+  @IsNumber()
+  credit: number;
```
Comment (yousinix) [inline src/users/dto/register-user.dto.ts]: `credit` is not sent in DTO, it defaults to zero in the prisma schema already.

File: src/users/interfaces/user.interface.ts [hunk 10]
```diff
@@ -1,7 +1,10 @@
-export interface User {
+export interface UserDto {
   id: string;
-  name: string;
-  key: string;
```
Comment (yousinix) [inline src/users/interfaces/user.interface.ts:4]: The `key` is sent back in this DTO, shouldn't be removed.

File: src/users/interfaces/user.interface.ts [hunk 11]
```diff
@@ -1,7 +1,10 @@
-export interface User {
+export interface UserDto {
```
Comment (yousinix) [inline src/users/interfaces/user.interface.ts]: Rename this file using DTO convention

File: src/users/users.service.ts [hunk 12]
```diff
@@ -1,39 +1,56 @@
...
-      key: null, // Don't save key in store
+  async login(loginUserDto: LoginUserDto): Promise<UserDto> {
+    const user = await this.prisma.user.findFirst({
+      where: {
+        id: loginUserDto.id,
+        password: loginUserDto.password,
+      },
     });
+    if (user === null) {
+      throw new NotFoundException('Incorrect user id or password.');
```
Comment (yousinix) [inline src/users/users.service.ts]: Should be `UnauthorizedException` (status code 401).

File: src/users/users.service.ts [hunk 13]
```diff
@@ -1,39 +1,56 @@
...
+    let user: UserDto;
+    try {
+      user = await this.prisma.user.create({
+        data: {
+          ...registerUserDto,
+          secret: encryptedSecret,
+        },
+      });
+    } catch (err) {
+      throw new NonUniqueException(err);
```
Comment (yousinix) [inline src/users/users.service.ts]: We should specify which field(s) are violating the uniqueness constraints using `IFieldValidationError`.

File: src/users/users.service.ts [hunk 14]
```diff
@@ -1,39 +1,56 @@
...
-      name: 'zero-based',
-      email: 'zerobasedteam@gmail.com',
-      secret,
-      key,
-    };
+    let user: UserDto;
+    try {
+      user = await this.prisma.user.create({
+        data: {
+          ...registerUserDto,
```
Comment (yousinix) [inline src/users/users.service.ts]: Passwords should be hashed.

**Review by yousinix — CHANGES_REQUESTED** [review]

File: src/users/users.controller.ts [hunk 15]
```diff
@@ -1,19 +1,30 @@
...
 @ApiTags('users')
 @Controller('users')
 export class UsersController {
   constructor(private readonly usersService: UsersService) {}
 
-  @ApiOperation({ summary: 'Creates a new user' })
+  @ApiOperation({ summary: 'Registers a new user.' })
   @ApiCreatedResponse({
     description: 'The user has been successfully created.',
   })
```
Comment (yousinix) [inline src/users/users.controller.ts]: We should specify all other paths/status codes too.

**Review by yousinix — CHANGES_REQUESTED** [review]

File: package.json [hunk 16]
```diff
@@ -34,6 +34,7 @@
     "class-transformer": "^0.4.0",
     "class-validator": "^0.13.1",
     "crypto-js": "^4.0.0",
+    "js-sha512": "^0.8.0",
```
Comment (MichaelSafwatHanna) [inline package.json]: we have [`crypto-js`](https://github.com/brix/crypto-js) in our dependencies, I checked their docs and they have hashing algorithms (including sha512), can you investigate using it? to minimize our dependency tree.

**Review by yousinix — DISMISSED** [review]

**Review by yousinix — APPROVED** [review]

**Pull request merged**